cmsplugin/cmsplugin.go

```diff
@@ -0,0 +1,480 @@
+	Options          []VoteOption `json:"options"`          // Vote option
+}
+
+// EncodeVote encodes Vote into a JSON byte slice.
+func EncodeVote(v Vote) ([]byte, error) {
+	return json.Marshal(v)
+}
+
+// DecodeVote decodes a JSON byte slice into a Vote.
+func DecodeVote(payload []byte) (*Vote, error) {
+	var v Vote
+
+	err := json.Unmarshal(payload, &v)
+	if err != nil {
+		return nil, err
+	}
+
+	return &v, nil
+}
+
+// CastVote is a signed vote.
+type CastVote struct {
+	Token     string `json:"token"`     // DCC ID
+	UserID    string `json:"publickey"` // User ID provided by cmswww
+	VoteBit   string `json:"votebit"`   // Vote bit that was selected, this is encode in hex
+	Signature string `json:"signature"` // Signature of the Token+VoteBit+UserID by the submitting user.
+}
+
+// EncodeCastVote encodes CastVotes into a JSON byte slice.
+func EncodeCastVote(cv CastVote) ([]byte, error) {
+	return json.Marshal(cv)
+}
+
+// DecodeCastVote decodes a JSON byte slice into a CastVote.
+func DecodeCastVote(payload []byte) (*CastVote, error) {
+	var cv CastVote
+
+	err := json.Unmarshal(payload, &cv)
+	if err != nil {
+		return nil, err
+	}
+
+	return &cv, nil
+}
+
+// CastVoteReply contains the signature or error to a cast vote command. The
+// Error and ErrorStatus fields will only be populated if something went wrong
+// while attempting to cast the vote.
+type CastVoteReply struct {
+	ClientSignature string       `json:"clientsignature"`       // Signature that was sent in
+	Signature       string       `json:"signature"`             // Signature of the ClientSignature
+	Error           string       `json:"error"`                 // Error status message
+	ErrorStatus     ErrorStatusT `json:"errorstatus,omitempty"` // Error status code
+}
+
+// EncodeCastVoteReply encodes CastVoteReply into a JSON byte slice.
+func EncodeCastVoteReply(cvr CastVoteReply) ([]byte, error) {
+	return json.Marshal(cvr)
+}
+
+// DecodeCastVoteReply decodes a JSON byte slice into a CastVote.
+func DecodeCastVoteReply(payload []byte) (*CastVoteReply, error) {
+	var cvr CastVoteReply
+
+	err := json.Unmarshal(payload, &cvr)
+	if err != nil {
+		return nil, err
+	}
+
+	return &cvr, nil
+}
+
+// UserWeight describes a single vote option.
+type UserWeight struct {
+	UserID string `json:"userid"` // Unique user id from cmswww.
+	Weight int64  `json:"weight"` // Calculated user voted weight, provided by cmswww.
+}
+
+const VersionStartVote = 1
+
+// StartVote instructs the plugin to commence voting on a proposal with the
+// provided vote bits.
+type StartVote struct {
+	// decred plugin only data
+	Version uint   `json:"version"` // Version of this structure
+	Payload string `json:"payload"` // JSON encoded StartVote
+	Token   string `json:"token"`   // Token
+
+	PublicKey   string       `json:"publickey"`   // Key used for signature.
+	UserWeights []UserWeight `json:"userweights"` // Array of User ID + weight
```

Comment:
You're saving the user weights on disk in both the StartVote and StartVoteReply mdstreams.  This is redundant.  Why do you need both?